src/rgw/rgw_rest_s3.cc

```diff
@@ -3749,12 +3761,80 @@ int RGW_Auth_S3::authorize_v4(RGWRados *store, struct req_state *s, bool force_b
 
   dout(10) << "credential scope = " << s->aws4_auth->credential_scope << dendl;
 
+  if (session_token) {
+    //Decode the session token
+    std::string decoded_str = rgw::from_base64(session_token);
+    std::map<std::string, std::string> decoded_token_map;
+
+    // For the time being the key will be part of ceph.conf
+    std::string key = std::move(store->ctx()->_conf->rgw_s3_auth_sts_key);
+
+    //Decrypt the session token
+    std::string decrypted_str;
+    auto* cryptohandler = s->cct->get_crypto_handler(CEPH_CRYPTO_AES);
+    if (! cryptohandler) {
+      return -EPERM;
+    }
+    bufferptr bp(key.c_str(), key.length());
+    int ret = cryptohandler->validate_secret(bp);
+    if (ret < 0) {
+      ldout(s->cct, 0) << "ERROR: Invalid secret key" << dendl;
+      return -EPERM;
+    }
+    string error;
+    auto* keyhandler = cryptohandler->get_key_handler(bp, error);
+    if (! keyhandler) {
+      return -EPERM;
+    }
+    error.clear();
+    bufferlist en_input, dec_output;
+    en_input.append(decoded_str);
+    ret = keyhandler->decrypt(en_input, dec_output, &error);
+    if (ret < 0) {
+      ldout(s->cct, 0) << "ERROR: Decryption failed: " << error << dendl;
+      return -EPERM;
+    } else {
+      dec_output.append('\0');
+      decrypted_str = dec_output.c_str();
+    }
+    //Extract useful info (like expiration time, secret key) from the token
+    boost::char_separator<char> sep("&");
+    boost::tokenizer<boost::char_separator<char>> tokens(decrypted_str, sep);
+    for (const auto& t : tokens) {
+      auto pos = t.find("=");
+      if (pos != string::npos) {
+        std::string key = t.substr(0, pos);
+        std::string value = t.substr(pos + 1, t.size() - 1);
+        decoded_token_map[key] = value;
+      }
+    }
+
+    //Check if the token has expired
+    if (decoded_token_map.find("expiration") != decoded_token_map.end()) {
+      std::string expiration = decoded_token_map["expiration"];
+      if (! expiration.empty()) {
+        const time_t exp = atoll(expiration.c_str());
+        time_t now;
+        time(&now);
+
+        if (now >= exp) {
+          return -EPERM;
+        }
+      }
+    }
+
+    if (decoded_token_map.find("secret_access_key") != decoded_token_map.end()) {
+      s->aws4_auth->secret_access_key = decoded_token_map["secret_access_key"];
+    }
+
+  } else {
   /* grab user information */
 
-  if (rgw_get_user_info_by_access_key(store, s->aws4_auth->access_key_id, *s->user) < 0) {
-    dout(10) << "error reading user info, uid=" << s->aws4_auth->access_key_id
-              << " can't authenticate" << dendl;
-    return -ERR_INVALID_ACCESS_KEY;
+    if (rgw_get_user_info_by_access_key(store, s->aws4_auth->access_key_id, *s->user) < 0) {
+      dout(10) << "error reading user info, uid=" << s->aws4_auth->access_key_id
+                << " can't authenticate" << dendl;
```

Comment:
@pritha-srivastava I believe would be good to use "can not" instead of can't.
I makes translation easy.
Your thoughts?